File: v1/examples/javascript/08-system-info.js
/**
 * Example 08: System & Server Information
 *
 * Demonstrates:
 *   - Server info (API version, product, license, timezone)
 *   - System info (OS, CPU, disk, network)
 *   - System health (CPU usage, memory, disk usage)
 *   - HDD S.M.A.R.T status
 *
 * NOTE: System info 'storage' parameter returns response field named 'disk'.
 */

const { NVR_API_KEY } = require('./config');
const { setApiKey, get } = require('./http');

async function main() {
  if (!NVR_API_KEY) {
    console.error('NVR_API_KEY environment variable is required');
    process.exit(1);
  }
  setApiKey(NVR_API_KEY);

  // ─────────────────────────────────────────────────
  // 1. Server Info (all at once)
  // ─────────────────────────────────────────────────
  console.log('=== Server Info ===');
  const infoRes = await get('/api/v1/info?all');

  if (infoRes.status === 200 && infoRes.body) {
    const info = infoRes.body;
    console.log(`  API Version: ${info.apiVersion}`);
    console.log(`  Site Name:   ${info.siteName}`);
    console.log(`  Product:     ${info.product?.name} v${info.product?.version}`);
    console.log(`  Timezone:    ${info.timezone?.name} (${info.timezone?.bias})`);
    console.log(`  License:     ${info.license?.type} (max ${info.license?.maxChannels} channels)`);
    console.log(`  User:        ${info.whoAmI?.uid} (${info.whoAmI?.name})`);
  }

  // ─────────────────────────────────────────────────
  // 2. System Info (individual items)
  //    Available items: os, cpu, storage, network
  //    NOTE: 'storage' request returns 'disk' field in response
  // ─────────────────────────────────────────────────
  console.log('\n=== System Info ===');

  // OS info
  const osRes = await get('/api/v1/system/info?item=os');
  if (osRes.status === 200 && osRes.body) {
    console.log(`  OS: ${JSON.stringify(osRes.body.os || osRes.body)}`);
  }

  // CPU info
  const cpuRes = await get('/api/v1/system/info?item=cpu');
  if (cpuRes.status === 200 && cpuRes.body) {
    console.log(`  CPU: ${JSON.stringify(cpuRes.body.cpu || cpuRes.body)}`);
  }

  // Storage info (response field is 'disk', not 'storage')
  const storageRes = await get('/api/v1/system/info?item=storage');
  if (storageRes.status === 200 && storageRes.body) {
    const disks = storageRes.body.disk || storageRes.body.storage || storageRes.body;
    console.log(`  Disk: ${JSON.stringify(disks)}`);
  }

  // Network info (includes lastUpdate field)
  const netRes = await get('/api/v1/system/info?item=network');
  if (netRes.status === 200 && netRes.body) {
    console.log(`  Network: ${JSON.stringify(netRes.body.network || netRes.body)}`);
    if (netRes.body.lastUpdate) {
      console.log(`  Last Update: ${netRes.body.lastUpdate}`);
    }
  }

  // Multiple items at once
  console.log('\n--- Multiple items ---');
  const multiRes = await get('/api/v1/system/info?item=os,cpu');
  if (multiRes.status === 200) {
    console.log(`  ${JSON.stringify(multiRes.body, null, 2)}`);
  }

  // ─────────────────────────────────────────────────
  // 3. System Health (real-time usage)
  //    Available items: cpu, memory, disk
  // ─────────────────────────────────────────────────
  console.log('\n=== System Health ===');
  const healthRes = await get('/api/v1/system/health');

  if (healthRes.status === 200 && healthRes.body) {
    const h = healthRes.body;

    if (Array.isArray(h.cpu)) {
      for (const c of h.cpu) {
        console.log(`  CPU Usage: ${c.usage?.total ?? 'N/A'}%`);
      }
    }
    if (h.memory) {
      const total = h.memory.totalPhysical || 0;
      const free = h.memory.freePhysical || 0;
      const used = total - free;
      const pct = total > 0 ? ((used / total) * 100).toFixed(1) : 'N/A';
      console.log(`  Memory: ${pct}% (${formatBytes(used)} / ${formatBytes(total)})`);
    }
    if (Array.isArray(h.disk)) {
      for (const d of h.disk) {
        const total = d.totalSpace || 0;
        const free = d.freeSpace || 0;
        const used = total - free;
        const pct = total > 0 ? ((used / total) * 100).toFixed(1) : 'N/A';
        console.log(`  Disk ${d.mount || ''}: ${pct}% (${formatBytes(used)} / ${formatBytes(total)})`);
      }
    }
  }

  // ─────────────────────────────────────────────────
  // 4. HDD S.M.A.R.T Status
  // ─────────────────────────────────────────────────
  console.log('\n=== HDD S.M.A.R.T ===');
  const smartRes = await get('/api/v1/system/hddsmart');
  console.log(`  Status: ${smartRes.status}`);

  if (smartRes.status === 200 && smartRes.body) {
    if (Array.isArray(smartRes.body)) {
      for (const disk of smartRes.body) {
        console.log(`  ${disk.model || disk.name || 'Disk'}: ${disk.health || disk.status || 'N/A'}`);
      }
    } else {
      console.log(`  ${JSON.stringify(smartRes.body, null, 2)}`);
    }
  }

}

function formatBytes(bytes) {
  if (!bytes) return '0 B';
  const units = ['B', 'KB', 'MB', 'GB', 'TB'];
  let i = 0;
  let val = bytes;
  while (val >= 1024 && i < units.length - 1) { val /= 1024; i++; }
  return `${val.toFixed(1)} ${units[i]}`;
}

main().catch(console.error);

File: v1/examples/swift/08-system-info.swift
/**
 * Example 08: System & Server Information
 *
 * Demonstrates:
 *   - Server info (API version, product, license, timezone)
 *   - System info (OS, CPU, disk, network)
 *   - System health (CPU usage, memory, disk usage)
 *   - HDD S.M.A.R.T status
 *
 * NOTE: System info 'storage' parameter returns response field named 'disk'.
 */

import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

func formatBytes(_ bytes: Double) -> String {
    guard bytes > 0 else { return "0 B" }
    let units = ["B", "KB", "MB", "GB", "TB"]
    var val = bytes
    var i = 0
    while val >= 1024 && i < units.count - 1 {
        val /= 1024
        i += 1
    }
    return String(format: "%.1f %@", val, units[i])
}

@main
struct App {
  static func main() async throws {
    let client = NvrClient()
    guard !client.apiKey.isEmpty else {
        print("NVR_API_KEY environment variable is required")
        exit(1)
    }
    client.setApiKey(client.apiKey)

    // -------------------------------------------------
    // 1. Server Info (all at once)
    // -------------------------------------------------
    print("=== Server Info ===")
    let infoRes = try await client.get("/api/v1/info?all")

    if infoRes.status == 200, let info = infoRes.body as? [String: Any] {
        let apiVersion = info["apiVersion"] ?? "N/A"
        let siteName = info["siteName"] ?? "N/A"
        print("  API Version: \(apiVersion)")
        print("  Site Name:   \(siteName)")

        if let product = info["product"] as? [String: Any] {
            let name = product["name"] ?? "N/A"
            let version = product["version"] ?? "N/A"
            print("  Product:     \(name) v\(version)")
        }
        if let tz = info["timezone"] as? [String: Any] {
            let name = tz["name"] ?? "N/A"
            let bias = tz["bias"] ?? "N/A"
            print("  Timezone:    \(name) (\(bias))")
        }
        if let license = info["license"] as? [String: Any] {
            let type = license["type"] ?? "N/A"
            let maxCh = license["maxChannels"] ?? "N/A"
            print("  License:     \(type) (max \(maxCh) channels)")
        }
        if let whoAmI = info["whoAmI"] as? [String: Any] {
            let uid = whoAmI["uid"] ?? "N/A"
            let name = whoAmI["name"] ?? "N/A"
            print("  User:        \(uid) (\(name))")
        }
    }

    // -------------------------------------------------
    // 2. System Info (individual items)
    //    Available items: os, cpu, storage, network
    //    NOTE: 'storage' request returns 'disk' field in response
    // -------------------------------------------------
    print("\n=== System Info ===")

    // OS info
    let osRes = try await client.get("/api/v1/system/info?item=os")
    if osRes.status == 200, let body = osRes.body as? [String: Any] {
        let osInfo = body["os"] ?? body
        print("  OS: \(client.toJsonString(osInfo as Any))")
    }

    // CPU info
    let cpuRes = try await client.get("/api/v1/system/info?item=cpu")
    if cpuRes.status == 200, let body = cpuRes.body as? [String: Any] {
        let cpuInfo = body["cpu"] ?? body
        print("  CPU: \(client.toJsonString(cpuInfo as Any))")
    }

    // Storage info (response field is 'disk', not 'storage')
    let storageRes = try await client.get("/api/v1/system/info?item=storage")
    if storageRes.status == 200, let body = storageRes.body as? [String: Any] {
        let disks = body["disk"] ?? body["storage"] ?? body
        print("  Disk: \(client.toJsonString(disks as Any))")
    }

    // Network info (includes lastUpdate field)
    let netRes = try await client.get("/api/v1/system/info?item=network")
    if netRes.status == 200, let body = netRes.body as? [String: Any] {
        let netInfo = body["network"] ?? body
        print("  Network: \(client.toJsonString(netInfo as Any))")
        if let lastUpdate = body["lastUpdate"] {
            print("  Last Update: \(lastUpdate)")
        }
    }

    // Multiple items at once
    print("\n--- Multiple items ---")
    let multiRes = try await client.get("/api/v1/system/info?item=os,cpu")
    if multiRes.status == 200, let body = multiRes.body {
        print("  \(client.toJsonString(body, pretty: true))")
    }

    // -------------------------------------------------
    // 3. System Health (real-time usage)
    //    Available items: cpu, memory, disk
    // -------------------------------------------------
    print("\n=== System Health ===")
    let healthRes = try await client.get("/api/v1/system/health")

    if healthRes.status == 200, let h = healthRes.body as? [String: Any] {
        if let cpuArr = h["cpu"] as? [[String: Any]] {
            for c in cpuArr {
                if let usage = c["usage"] as? [String: Any], let total = usage["total"] {
                    print("  CPU Usage: \(total)%")
                }
            }
        }
        if let memory = h["memory"] as? [String: Any] {
            let total = memory["totalPhysical"] as? Double ?? 0
            let free = memory["freePhysical"] as? Double ?? 0
            let used = total - free
            let pct = total > 0 ? String(format: "%.1f", (used / total) * 100) : "N/A"
            print("  Memory: \(pct)% (\(formatBytes(used)) / \(formatBytes(total)))")
        }
        if let diskArr = h["disk"] as? [[String: Any]] {
            for d in diskArr {
                let total = d["totalSpace"] as? Double ?? 0
                let free = d["freeSpace"] as? Double ?? 0
                let used = total - free
                let pct = total > 0 ? String(format: "%.1f", (used / total) * 100) : "N/A"
                let mount = d["mount"] as? String ?? ""
                print("  Disk \(mount): \(pct)% (\(formatBytes(used)) / \(formatBytes(total)))")
            }
        }
    }

    // -------------------------------------------------
    // 4. HDD S.M.A.R.T Status
    // -------------------------------------------------
    print("\n=== HDD S.M.A.R.T ===")
    let smartRes = try await client.get("/api/v1/system/hddsmart")
    print("  Status: \(smartRes.status)")

    if smartRes.status == 200, let body = smartRes.body {
        if let disks = body as? [[String: Any]] {
            for disk in disks {
                let model = disk["model"] as? String ?? disk["name"] as? String ?? "Disk"
                let health = disk["health"] as? String ?? disk["status"] as? String ?? "N/A"
                print("  \(model): \(health)")
            }
        } else {
            print("  \(client.toJsonString(body, pretty: true))")
        }
    }
  }
}
